app/presenters/application_presenter.rb

```diff
@@ -0,0 +1,115 @@
+# frozen_string_literal: true
+
+class ApplicationPresenter < SimpleDelegator
+  include ActionView::Helpers::FormTagHelper
+  include ActionView::Helpers::UrlHelper
+  include Rails.application.routes.url_helpers
+
+  attr_reader :action, :controller
+
+  # Master list of fields that could be displayed on any page.
+  ALL_FIELDS = %i[].freeze
+
+  # Not displayed on the show page
+  HIDDEN_FIELDS = %i[].freeze
+
+  # Displayed on the edit page but not editable
+  READ_ONLY_FIELDS = %i[].freeze
+
+  # Only editable when creating a new object
+  READ_ONLY_AFTER_PERSISTED_FIELDS = %i[].freeze
+
+  # Default character size of editable fields
+  FIELD_SIZE = 60
+
+  def self.attribute_scope
+    @attribute_scope ||= "activerecord.attributes." + name.sub(%r{Presenter$}, "").underscore
+  end
+
+  def self.value_scope
+    @value_scope ||= name.sub(%r{Presenter$}, "").underscore + ".values"
+  end
+
+  def self.field_label(field)
+    I18n.t(field, scope: attribute_scope)
+  end
+
+  def initialize(obj, controller: nil, action: :show)
+    super obj
+    @controller = controller
+    @action = action
+  end
+
+  # By default form cancel button goes to show page if persisted,
+  # index page otherwise.
+  def cancel_path
```

Comment:
Not sure the metaprogramming/class name manipulation is worth the DRY savings here.

Reply:
It's nightmarish to read, but if it saves someone typing mind-numbing boilerplate...

Reply:
I think fine to keep for now -- if we run into strange issues debugging stuff or get ourselves confused in the future we might find that we would rather get rid of this and just have the boilerplate, e.g.:

```ruby
class ApplicationPresenter
  def cancel_path
    if persisted?
      send show_route_helper, self
    else
      send index_route_helper
    end
   end

end

class WhateverPresenter < ApplicationPresenter
  def show_route_helper
    :whatever_path
    # or maybe :whatever_path.to_proc
  end

  def index_route_helper
    :whatevers_path
  end
end
```

There might be some way to use `url_for` to get those urls out. I have often wished there was something defined on the model that gave you the name of the route helpers, but I've never been able to find it.

In any case, I'm fine with this as-is for now, and hopefully we won't have to touch it..?

Reply:
I've been navigating back and forth checking the registrations in particular, sending mails and checking the landing pages, no issues there as far as using the cancel button for navigation. Hoping we don't have to touch it for a while. I'm gonna leave it in place.